modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java

```diff
@@ -0,0 +1,340 @@
+                                    new Object[] {env, granule.getName()});
+                        }
+                        allBounds.add(env);
+                    }
+                }
+            } catch (Exception e) {
+                String message = "Failed to get the bounds on " + granule.getName();
+                if (getStore().isTolerant()) {
+                    LOGGER.log(Level.WARNING, message, e);
+                    return null;
+                } else {
+                    throw new IOException(message, e);
+                }
+            }
+        }
+        ReferencedEnvelope result =
+                new ReferencedEnvelope(getSchema().getCoordinateReferenceSystem());
+        for (ReferencedEnvelope bound : allBounds) {
+            Envelope env = new Envelope(bound);
+            result.expandToInclude(env);
+        }
+        if (result.isNull()) {
+            return null;
+        } else {
+            return result;
+        }
+    }
+
+    @Override
+    protected int getCountInternal(Query query) throws IOException {
+        List<Long> counts = new ArrayList<>();
+        for (VectorMosaicGranule granule : config.getGranules()) {
+            DataStore dataStore = granule.getDataStore();
+            if (dataStore != null) {
+                SimpleFeatureSource source =
+                        dataStore.getFeatureSource(granule.getGranuleTypeName());
+                Query q = new Query(query);
+                q.setTypeName(granule.getGranuleTypeName());
+                Long count = Long.valueOf(source.getCount(q));
+                counts.add(count);
+            }
+        }
+        // aggregate the counts
+        long total = 0;
+        for (Long count : counts) {
+            try {
+                if (count > 0) {
+                    total += count;
+                } else {
+                    // one of the sources found it was too costly to count
+                    return -1;
+                }
+            } catch (Exception e) {
+                throw new IOException("Failed to count on a delegate store", e);
+            }
+        }
+        return (int) total;
+    }
+
+    @Override
+    protected FeatureReader<SimpleFeatureType, SimpleFeature> getReaderInternal(Query query)
+            throws IOException {
+        try {
+            if (config.getGranules() == null || config.getGranules().isEmpty()) {
+                throw new IOException("No granules found for mosaic");
+            }
+            // get the first store
+            VectorMosaicGranule firstGranule = config.getGranules().get(0);
+            Query psQuery = new Query(query);
+            psQuery.setTypeName(firstGranule.getGranuleTypeName());
+            DataStore ps = firstGranule.getDataStore();
+            SimpleFeatureType target = DataUtilities.createView(ps, psQuery).getSchema();
+            target = retypeNameSchema(target);
+
+            FeatureQueue queue = new FeatureQueue();
```

Comment:
Nope, don't accumulate features in memory, it's gonna cause a OOM (this store is for large datasets).
Instead, create a reader that has two delegate readers (or feature iterators):
* One over the index table, pointing at the current granule
* One over the  current granule
*  When the current granule iterator is done, close, skip to the next granule, open a new iterator, and move on.
This keeps at most one feature in memory. Similar (but not same) to this one:
https://github.com/geoserver/geoserver/blob/9e40cbbccb9626b8d08516c8eacaf5a8f168189b/src/extension/csw/api/src/main/java/org/geoserver/csw/feature/CompositeIterator.java

Reply:
Fixed